finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/StreamProxy.scala

```diff
@@ -0,0 +1,34 @@
+package com.twitter.finagle.buoyant.h2
+import com.twitter.util.{Future, Try}
+
+/**
+ * Wraps an underlying [[Stream]] with an [[StreamProxy.onFrame onFrame]] function.
+ * The `onFrame` function is called for each frame in the stream.
+ * @param underlying the [[Stream]] wrapped by this proxy
+ * @param onFrame function called for each [[Frame]] in the underlying [[Stream]]
+ */
+class StreamProxy(underlying: Stream, onFrame: Try[Frame] => Unit) extends Stream {
+  override def isEmpty: Boolean = underlying.isEmpty
+
+  override def read(): Future[Frame] = underlying.read().respond(frame => onFrame(frame))
```

Comment:
It was point-free for a while & changed because I thought it needed to free the frame after, realized that was wrong, and this never got changed back :)